org.eclipse.scanning.api/src/org/eclipse/scanning/api/device/models/DeviceWatchdogModel.java

```diff
@@ -1,219 +1,219 @@
+    
+    1. Mode is normal (8)
+    2. t > c
+    3. t < (p-Tf)-w
+    
+    </pre>
+    
+<h3>Ring Mode</h3>
+
+The "Ring Mode" PV is SR-CS-RING-01:MODE.
+
+This PV has various states: 
+<img src="./doc/modes.png" /> 
+
+In brief though, the only one you need to care about is state 8 = VMX. This is "normal" mode now that we've installed the new VMX (AKA DDBA) components.
+
+If this PV = 8, then we're in normal mode. If this PV is anything else, then we're in some other state.
+
+ * @author Matthew Gerring
+ *
+ */
+public class DeviceWatchdogModel {
+
+	/**
+	 * This is an expression of scannables which are monitored and the expression
+	 * reevaluated if they change during a scan. Each variable in the scan *must*
+	 * be a scannable name. The expression will be evaluated using JEXL which is
+	 * available through the IExpressionService. Any expression
+	 */
+	private String expression; // e.g. 'beamcurrent >= 1.0 && !portshutter.equalsIgnoreCase("Closed")'
+	private String message;
+	
+	// t
+	private String countdownName; // e.g. "topup", "countdown" PV likely to be SR-CS-FILL-01:COUNTDOWN which is in s
+	
+	// c in ms
+	private long   cooloff;       // time in ms before topup for which the scan should be paused.
+	
+	// w in ms
+	private long   warmup;        // time in ms after topup the scan should wait before starting.
+	
+	// p in ms
+	private long   period = 10*60*1000; // Period in ms, default is 10min
+	
+	// Tf in ms
+	private long   topupTime = 15*1000; // The time that a topup takes. This is varible but in normal mode <= 15s
+	
+	// The name of the mode pv, if any. 
+	private String modeName;            // If this is set the PV will be checked to ensure that the topup mode is as expected.
+	
+	public String getCountdownName() {
+		return countdownName;
+	}
+	public void setCountdownName(String monitorName) {
+		this.countdownName = monitorName;
+	}
+	public long getCooloff() {
+		return cooloff;
+	}
+	public void setCooloff(long cooloff) {
+		this.cooloff = cooloff;
+	}
+	public long getWarmup() {
+		return warmup;
+	}
+	public void setWarmup(long warmup) {
+		this.warmup = warmup;
+	}
+	public String getMessage() {
+		return message;
+	}
+	public void setMessage(String message) {
+		this.message = message;
+	}
+	public String getExpression() {
+		return expression;
+	}
+	public void setExpression(String expression) {
+		this.expression = expression;
+	}
+	@Override
+	public int hashCode() {
+		final int prime = 31;
+		int result = 1;
+		result = prime * result + (int) (cooloff ^ (cooloff >>> 32));
+		result = prime * result + ((countdownName == null) ? 0 : countdownName.hashCode());
+		result = prime * result + ((expression == null) ? 0 : expression.hashCode());
+		result = prime * result + ((message == null) ? 0 : message.hashCode());
+		result = prime * result + (int) (warmup ^ (warmup >>> 32));
+		return result;
+	}
+	@Override
+	public boolean equals(Object obj) {
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this method to reduce its Cognitive Complexity from 17 to the 15 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)